master/internal/api_resourcepool.go

```diff
@@ -169,6 +169,29 @@ func (a *apiServer) ListWorkspacesBoundToRP(
 	}, nil
 }
 
+func (a *apiServer) GetUnboundResourcePools(ctx context.Context,
+	req *apiv1.GetUnboundResourcePoolsRequest,
+) (*apiv1.GetUnboundResourcePoolsResponse, error) {
+	getPoolsResp, err := a.m.rm.GetResourcePools(a.m.system, &apiv1.GetResourcePoolsRequest{})
+	if err != nil {
+		return nil, err
+	}
+
+	var poolNames []string
+	for _, pool := range getPoolsResp.ResourcePools {
+		poolNames = append(poolNames, pool.Name)
+	}
+
+	unboundPools, err := db.GetUnboundRPs(ctx, poolNames)
+	if err != nil {
+		return nil, err
+	}
+
+	resp := apiv1.GetUnboundResourcePoolsResponse{ResourcePools: unboundPools}
+
+	return &resp, a.paginate(&resp.Pagination, &resp.ResourcePools, req.Offset, req.Limit)
```

Comment:
Not sure I understand this. Is this equivalent to:
```
return &apiv1.GetUnboundResourcePoolsResponse{
		ResourcePools: resourcePools, 
                Pagination: pagination,
 	}, nil
```

Reply:
Sorta, yeah. Unfortunately, [`paginate`](https://github.com/determined-ai/determined/blob/main/master/internal/api.go#L35) is both very complicated and has an inaccurate doc comment. It actually mutates the first two arguments you pass in, somewhat like [`json.Marshal`](https://pkg.go.dev/encoding/json#Marshal), by slicing the paginated values and filling in the pagination data (offset, limit, total, start index, and end index).

IMO this is made even more obtuse when it's used like this (combining a `return` line and the call to the mutating function). To be fair to Eric, this is done a lot in the code, so he was following established idioms. But I think it'd be more clear if the two were separated, though that'd make for an extra line of code.

```go
resp := apiv1.GetUnboundResourcePoolsResponse{ResourcePools: unboundPools}
err = a.paginate(&resp.Pagination, &resp.ResourcePools, req.Offset, req.Limit)

return &resp, err
```

Reply:
`a.paginate` paginates the response for you given the parameters in the request

Reply:
It only returns an error. It mutates parameters *rather than* returning results, which is atypical and usually worth documenting—which is not done in this case.

Reply:
I can refactor to the longer version if it helps with readability

Reply:
I think that'd be an improvement, though I'd categorize it as a nit.